src/Lean/Data/LParsec.lean

```diff
@@ -0,0 +1,138 @@
+
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Dany Fabian
+-/
+
+namespace Lean
+
+namespace LParsec
+
+inductive ParseResult (α : Type) where
+  | success (pos : String.Iterator) (res : α)
+  | error (pos : String.Iterator) (err : String)
+  deriving Repr
+
+open ParseResult
+
+def Parser (α : Type) : Type := String.Iterator → ParseResult α
+
+instance (α : Type) : Inhabited (Parser α) :=
+  ⟨λ it => error it ""⟩
+
+@[inline]
+protected def pure (a : α) : Parser α := λ it => 
+ success it a 
+ 
+@[inline]
+def bind {α β : Type} (f : Parser α) (g : α → Parser β) : Parser β := λ it =>
+  match f it with
+  | success rem a => g a rem
+  | error pos msg => error pos msg
+  
+@[inline]
+instance : Monad Parser :=
+  { pure := LParsec.pure, bind }
+  
+@[inline]
+def fail (msg : String) : Parser α := fun it =>
+  error it msg
+
+@[inline]
+def orElse (p q : Parser α) : Parser α := fun it =>
```

Comment:
Note that this definitions is different from the classic parsec definition: the parsec orelse tries the second parser *only if the first one didn't consume any input.*

I believe this definition here could lead to some extremely unexpected performance pitfalls.  For example, take a valid XML document and add `<` to the end.  Then a top-level orElse will parse the document multiple times (because it backtracks after parsing the whole document).

Reply:
good point, thanks for flagging it up. This is a pure oversight, not intentional.